src/main/java/net/snowflake/client/core/SecureStorageLinuxManager.java

```diff
@@ -43,58 +44,64 @@ private SecureStorageLinuxManager() {
+    ObjectNode res = mapper.createObjectNode();
+    for (Map.Entry<String, Map<String, String>> elem : localCredCache.entrySet()) {
+      String elemHost = elem.getKey();
+      Map<String, String> hostMap = elem.getValue();
+      ObjectNode hostNode = mapper.createObjectNode();
+      for (Map.Entry<String, String> elem0 : hostMap.entrySet()) {
+        hostNode.put(elem0.getKey(), elem0.getValue());
+      }
+      res.set(elemHost, hostNode);
+    }
+    return res;
+  }
+
+  public synchronized SecureStorageStatus setCredential(
+      String host, String user, String type, String token) {
     if (Strings.isNullOrEmpty(token)) {
       logger.info("No token provided");
       return SecureStorageStatus.SUCCESS;
     }
 
-    idTokenCache.computeIfAbsent(host.toUpperCase(), newMap -> new HashMap<>());
+    localCredCache.computeIfAbsent(host.toUpperCase(), newMap -> new HashMap<>());
 
-    Map<String, String> currentUserMap = idTokenCache.get(host.toUpperCase());
-    currentUserMap.put(user.toUpperCase(), token);
-
-    ObjectNode out = mapper.createObjectNode();
-    for (Map.Entry<String, Map<String, String>> elem : idTokenCache.entrySet()) {
-      String elemHost = elem.getKey();
-      Map<String, String> userMap = elem.getValue();
-      ObjectNode userNode = mapper.createObjectNode();
-      for (Map.Entry<String, String> elem0 : userMap.entrySet()) {
-        userNode.put(elem0.getKey(), elem0.getValue());
-      }
-      out.set(elemHost, userNode);
-    }
-    fileCacheManager.writeCacheFile(out);
+    Map<String, String> hostMap = localCredCache.get(host.toUpperCase());
+    hostMap.put(buildCredName(user, type), token);
 
+    fileCacheManager.writeCacheFile(localCacheToJson());
     return SecureStorageStatus.SUCCESS;
   }
 
-  public String getCredential(String host, String user) {
+  public synchronized String getCredential(String host, String user, String type) {
     JsonNode res = fileCacheManager.readCacheFile();
     readJsonStoreCache(res);
 
-    Map<String, String> userMap = idTokenCache.get(host.toUpperCase());
+    Map<String, String> hostMap = localCredCache.get(host.toUpperCase());
 
-    if (userMap == null) {
+    if (hostMap == null) {
       return null;
     }
 
-    return userMap.get(user.toUpperCase());
+    return hostMap.get(buildCredName(user, type));
   }
 
-  /**
-   * Since Linux doesn't have secure local storage right now, this function's input parameters are
-   * only here to be consistent with Mac/Win platform. We don't really use these parameters. If in
-   * the future, deletion for a specific credential is needed, we can change this function to
-   * satisfy it."
-   */
-  public SecureStorageStatus deleteCredential(String host, String user) {
-    fileCacheManager.deleteCacheFile();
-    idTokenCache.clear();
+  /** May delete credentials which doesn't belong to this process */
+  public synchronized SecureStorageStatus deleteCredential(String host, String user, String type) {
```

Comment:
Because there will be more than one type of credentials in local file cache on Linux, we don't simple delete the cache file. Partially deletion is supported now.